sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/AcrClientCredentials.cs

```diff
@@ -0,0 +1,309 @@
+        /// Acquires the required scope for a specific operation. This will be done by obtaining a challenge and parsing out the scope
+        /// from the ww-Authenticate header. In the event of failure (Some endpoints do not seem to return the scope) it will attempt
+        /// resolution through a small local resolver.
+        /// <param name='scope'> The scope for the particuar operation. Can be obtained from the Www-Authenticate header.
+        /// </summary>
+
+        public async Task<string> getScope(string operation, string method, string path)
+        {
+
+            if (_acrScopes.ContainsKey(method + ">" + operation))
+            {
+                return _acrScopes[method + ">" + operation];
+            }
+
+            HttpClient runtimeClient = new HttpClient();
+            HttpResponseMessage response = null;
+            string scope;
+            try
+            {
+                response = await runtimeClient.SendAsync(new HttpRequestMessage(new HttpMethod(method), operation));
+                Dictionary<string, string> data = parseHeader(response.Headers.GetValues("Www-Authenticate").FirstOrDefault());
+                scope = data.ContainsKey("scope") ? data["scope"] : hardcodedScopes(path);
+                _acrScopes[method + ">" + operation] = scope;
+            }
+            catch (Exception e)
+            {
+                throw new Exception("Could not identify appropriate Token scope: " + e.Message);
+            }
+            return scope;
+
+        }
+
+        /// <summary>
+        /// Local resolver for endpoints that will often return no scope.
+        /// <param name='operation'> Operation for which a scope is necessary
+        /// </summary>
+        private string hardcodedScopes(string operation)
+        {
+            switch (operation)
+            {
+                case "/acr/v1/_catalog":
+                case "/v2/":
+                    return "registry:catalog:*";
+                default:
+                    throw new Exception("Could not determine appropriate scope for the specified operation");
+
+            }
+        }
+
+        /// <summary>
+        /// Meant to parse out comma separated key value pairs delineated by the = sign. Accepts strings of the
+        /// format "key=value,key2=value2..." . Note this method is meant to provide limited functionality and
+        /// is not very robust.
+        ///  </summary>
+        private Dictionary<string, string> parseHeader(string header)
+        {
+
+            Dictionary<string, string> parsed = new Dictionary<string, string>();
+
+            Regex re = new Regex(@"([\w\s]+)=""[^""]+"""); //Regex is required to allow multiple scopes like push,pull
+            MatchCollection parts = re.Matches(header);
+
+            foreach (Match part in parts)
+            {
+                string[] keyValues = part.ToString().Split('=');
+                parsed.Add(keyValues[0], headerTrim(keyValues[1]));
```

Comment:
Can `keyValues` have less than 2 elements?

Reply:
It shouldn't be possible. The regex requires that the string have an '=' in it. And the array returned by String.Split('=') will have a second element in it.

That being said I personally tend to lean away from regexes, so I might change this method to not use a regex.